pkg/binding/crd.go

```diff
@@ -35,19 +37,50 @@ type APIResource struct {
 }
 
 // Handle CRDs should account for CRDs being added or deleted to start/stop new informers as needed
-func (c *Controller) handleCRD(_ runtime.Object) error {
-	toStartList, toStopList, err := c.checkAPIResourcesForUpdates()
-	if err != nil {
-		return err
+func (c *Controller) handleCRD(obj runtime.Object) error {
+	uObj := obj.(*unstructured.Unstructured)
+	var crdObj *apiextensionsv1.CustomResourceDefinition
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(uObj.UnstructuredContent(), &crdObj); err != nil {
+		return fmt.Errorf("failed to convert Unstructured to CRD: %w", err)
+	}
+
+	// for each CustomResourceDefinitionVersion, follow a decision tree to tell whether the corresponding gvr should be watched
+	toStartList, toStopList := []APIResource{}, []string{}
+	for _, ver := range crdObj.Spec.Versions {
+		gvr := APIResource{
+			groupVersion: schema.GroupVersion{
+				Group:   crdObj.Spec.Group,
+				Version: ver.Name,
+			},
+			resource: metav1.APIResource{
+				Name: crdObj.Spec.Names.Plural,
+				Kind: crdObj.Spec.Names.Kind,
+			},
+		}
+		key := util.KeyForGroupVersionKind(gvr.groupVersion.Group, gvr.groupVersion.Version, gvr.resource.Kind)
+		if isBeingDeleted(obj) {
+			toStopList = append(toStopList, key)
+			continue
+		}
+		if !c.includedToWatch(gvr) {
+			toStopList = append(toStopList, key)
+			continue
+		}
+		if !ver.Served {
+			toStopList = append(toStopList, key)
+			continue
+		}
+		toStartList = append(toStartList, gvr)
 	}
 
 	go c.startInformersForNewAPIResources(toStartList)
 
 	for _, key := range toStopList {
-		c.logger.Info("API removed, stopping informer.", "key", key)
-		stopper := c.stoppers[key]
-		// close channel
-		close(stopper)
+		c.logger.Info("API should be removed, stopping informer.", "key", key)
```

Comment:
FYI, I would let the log message carry the information from `ok` below.

Reply:
Added in latest push.

Reply:
I do not see that change in https://github.com/kubestellar/kubestellar/blob/48f75f51b37bd325ad4c075aa15ad0c6b2158d6b/pkg/binding/crd.go#L84-L85

Reply:
FYI: I was trying to suggest that there would be just one log message in both cases (`ok` and `!ok`), with the log messages somehow letting the reader know whether `ok` was true or not. For example, have two different log statements (in the two arms of the `if`), which would not need an additional log statement up front.